IOperation Test Hook Pt 1

333fred opened the pull request:
Followup to https://github.com/dotnet/roslyn/pull/49534. That PR is already unicorning, so I'm going to split it up into a few smaller series of PRs. For simplicity of review, I'm going to take the following strategy: I'll cherry-pick the original commit to this branch, then create a separate commit to address any comments on the original commit right on top of it. I'll also copy over the relevant comment from the original commit and link to the original, then answer as a separate reply.

333fred commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IObjectCreationExpression.cs:
```diff
@@ -1623,7 +2360,75 @@ class C2 : IEnumerable<int>
                         OutConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
 ";
-            VerifyOperationTreeAndDiagnosticsForTest<InitializerExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics, parseOptions: ImplicitObjectCreationOptions);
+            VerifyOperationTreeAndDiagnosticsForTest<InitializerExpressionSyntax>(comp, expectedOperationTree, expectedDiagnostics);
```
Aleksey commented:

> It looks like all tests adjusted in the second commit are using target-typed new. I am assuming regular `new` had the same issue. Please confirm.
> It would be good to add verification for a couple of interesting scenarios with a regular `new`.

333fred replied:
It looks like it doesn't, because the existing code uses a syntax check rather than IOperation check, which I missed on my first pass. I'll update the syntax check instead, we already have coverage of these scenarios for regular object creation.

---
In reply to: [551544303](https://github.com/dotnet/roslyn/pull/50111#discussion_r551544303) [](ancestors = 551544303)

333fred commented on src/Compilers/Test/Core/Compilation/TestOperationVisitor.cs:
```diff
@@ -578,7 +578,8 @@ private void VisitMemberReference(IMemberReferenceOperation operation, IEnumerab
                     operation.Instance is IInstanceReferenceOperation)
                 {
-                    Assert.False(operation.Instance.IsImplicit, $"Implicit {nameof(IInstanceReferenceOperation)} on {operation.Syntax}");
+                    Assert.True(!operation.Instance.IsImplicit || operation.Instance.HasErrors(operation.SemanticModel.Compilation),
```
Aleksey commented:

> It feels like this breaks one of the invariants that we have. Why the original condition was fine for a regular member access and a problem for patterns?
> It feels to me that we should fix the product instead and maintain the invariant.

333fred replied:
> Why the original condition was fine for a regular member access and a problem for patterns?

Because the implementation included a receiver unconditionally. I'll adjust the implementation to not do so and restore this.

---
In reply to: [551545224](https://github.com/dotnet/roslyn/pull/50111#discussion_r551545224) [](ancestors = 551545224)

333fred commented on src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IIsPatternExpression.cs:
```diff
@@ -1455,6 +1457,12 @@ class D
+            var isPattern = tree.GetRoot().DescendantNodes().OfType<IsPatternExpressionSyntax>().Single();
+            var model = compilation.GetSemanticModel(tree);
+            var instanceReceiver = model.GetOperation(isPattern).Descendants().OfType<IInstanceReferenceOperation>().Single();
+            TestOperationVisitor.Singleton.Visit(instanceReceiver);
```
Aleksey commented:

> It is not clear to me what is this doing which VerifyOperationTreeForTest wouldn't already do. Could you please elaborate?

333fred replied:
`VerifyOperationTreeForTest` does not call the `TestOperationVisitor` on operation nodes. Perhaps it should, but that is what this is accomplishing.

---
In reply to: [551558624](https://github.com/dotnet/roslyn/pull/50111#discussion_r551558624) [](ancestors = 551558624)

333fred commented on src/Compilers/CSharp/Portable/Symbols/Tuples/TupleErrorFieldSymbol.cs:
```diff
@@ -177,5 +177,24 @@ public bool Equals(TupleErrorFieldSymbol other, TypeCompareKind compareKind)
                 TypeSymbol.Equals(ContainingType, other.ContainingType, compareKind);
         }
+
+        internal override FieldSymbol AsMember(NamedTypeSymbol newOwner)
```
Aleksey commented:

> I think we want targeted tests for this implementation to cover breadth of possible scenarios, including long tuples, etc. We should verify internal state (all the fields) of the symbol we get back.

333fred commented on src/Compilers/CSharp/Portable/Symbols/Tuples/TupleErrorFieldSymbol.cs:
>default [](start = 46, length = 7)

Aleksey commented:

> This doesn't feel right.

333fred commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
>indexes [](start = 45, length = 7)

Aleksey commented:

> I am not sure what "indexes" is referring to. Could you please elaborate?

333fred commented on src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:
```diff
@@ -3023,6 +3023,8 @@ private static bool AreCloseEnough(Symbol original, Symbol updated)
             {
                 (LambdaSymbol l, NamedTypeSymbol n) _ when n.IsDelegateType() => AreLambdaAndNewDelegateSimilar(l, n),
                 (FieldSymbol { ContainingType: { IsTupleType: true }, TupleElementIndex: var oi } originalField, FieldSymbol { ContainingType: { IsTupleType: true }, TupleElementIndex: var ui } updatedField) =>
+                    // For tuples with their indexes being private fields, such as in `InferredName_ConditionalOperator_UseSite`, when we do reinference we don't end up inferring the resulting type correctly,
```
Added a more elaborate comment.

---
In reply to: [551560779](https://github.com/dotnet/roslyn/pull/50111#discussion_r551560779) [](ancestors = 551560779)

333fred replied:
Actually, this comment can just removed now that `AsMember` is correctly handling tuple element indexes for error fields.

---
In reply to: [551560877](https://github.com/dotnet/roslyn/pull/50111#discussion_r551560877) [](ancestors = 551560877,551560779)

333fred commented:
@AlekseyTs @dotnet/roslyn-compiler for review.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -27408,5 +27408,66 @@ void symbolValidator(ModuleSymbol module)
+                AssertEx.Equal($"System.Object{nullabilityString} System.ValueTuple<System.Object!, System.Object?, System.Object!, System.Object?, System.Object!, System.Object?, System.Object!, (System.Object?, System.Object!)>.Item{i + 1}",
+                               newField.CorrespondingTupleField.ToTestDisplayString(includeNonNullable: true));
+
+                Assert.Equal(i, newField.TupleElementIndex);
```
>Assert.Equal(i, newField.TupleElementIndex); [](start = 16, length = 44)

Consider asserting that original field has the same syntax.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -27408,5 +27408,66 @@ void symbolValidator(ModuleSymbol module)
         }
+
+        [Fact]
+        public void TupleAsMemberTests()
```
>TupleAsMemberTests [](start = 20, length = 18)

Do we cover scenarios of ```IsDefaultTupleElement``` true and false?

AlekseyTs commented:
Done with review pass (commit 7)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [754817687](https://github.com/dotnet/roslyn/pull/50111#issuecomment-754817687) [](ancestors = 754817687)

AlekseyTs reviewed (approved):
LGTM (commit 9)

333fred commented:
@dotnet/roslyn-compiler for a second review.

jcouv commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -2292,20 +2292,17 @@ internal IOperation CreatePropertySubpatternMember(Symbol? symbol, ITypeSymbol m
         {
             var nameSyntax = (syntax is SubpatternSyntax subpatSyntax ? subpatSyntax.NameColon?.Name : null) ?? syntax;
             bool isImplicit = nameSyntax == syntax;
+            var receiver = symbol?.IsStatic == false ? new InstanceReferenceOperation(InstanceReferenceKind.PatternInput, _semanticModel, nameSyntax, matchedType, isImplicit: true) : null;
```
>receiver [](start = 16, length = 8)

Can we extract this to a local function instead? We don't need to create a receiver in the `default:` case.

jcouv reviewed (approved):
LGTM Thanks (iteration 9) with one suggestion

333fred commented:
I've rebased and squashed the fixup commits to keep history neater.
